### DISCUSSION TITLE
Blocking training rows to avoid overfitting (custom bootstrap) - multivariate outcomes regression

### DISCUSSION THREAD

**Discussion by thengl** [kogalur/randomForestSRC#331] [Q&A]
Many thanks for your package and all support. I am now using the pkg more intensively. My especial interest is using the [multivariate outcomes regression](https://opengeohub.github.io/spatial-prediction-eml/spatial-interpolation-in-3d-using-ensemble-ml.html#modeling-multiple-geochemicals-using-a-single-ml-model) and here I have 2 questions:

1. I would like to block rows that are spatially clustered to prevent [overfitting](https://opengeohub.github.io/spatial-sampling-ml/resampling-methods-for-machine-learning.html#testing-mapping-accuracy-using-resampling-and-blocking). I am currently using your function for this `randomForestSRC:::make.sample`, which makes sure that all 0 and all 1 values selected belong to different site / spatial blocks. Maybe I have missed this, but you do not provide blocking option i.e. some class column or similar that user can specify which points can go either for training or for bootstraping?

2.  In the case of the multivariate outcomes regression is it possible to derive prediction errors for each new prediction location and how? I am thinking about using [`quantreg` function](https://www.randomforestsrc.org/reference/quantreg.rfsrc.html) but maybe this can not be combined with multivariate outcomes?

**Comment by ishwaran** [comment]
1. I'm not totally clear about what you are asking here.  The `make.sample` function creates a 0/1 matrix with `ntree` columns and `n` rows (where `n` is the sample size) and each column signifies for that specific tree what the in-sample (inbag) data will be for that tree by using a 1 and what the out-of-bag (OOB) data will be for that tree by using a 0. 
2. Again, I'm not clear about what you are asking.  Maybe this discussion is helpful?  https://github.com/kogalur/randomForestSRC/discussions/284#discussioncomment-3070462

**Reply by thengl** [reply]
I think 1 is resolved. I just wanted to check if my code was correct. Thank you!
For 2 the preferred thing would that during `predict` one can always request to produce predictions of the mean + prediction error / lower-upper quantiles. See e.g. https://cran.r-project.org/package=forestError.

**Comment by ishwaran** [comment]
Then the solution I presented in 2 will work.  Let me repeat the salient parts here.

Here we obtain OOB residuals from a bivariate regression
```
## fit bivariate forest, extract residual
o <- rfsrc(cbind(mpg, disp) ~ ., mtcars)
yhat <-  get.mv.predicted(o)
res <- o$yvar - yhat

```

You can then obtain quantiles of the residuals to obtain quantiles for your estimated value.  This is a very effective method for producing prediction intervals and is the default method used by `quantreg` by the package.

I did not look in detail at the package you sent (they did not provide a reference to any paper, so not easy to determine what is going on under the hood there) but I believe this is what they are doing essentially.

**Reply by thengl** [reply]
And how do you recommend producing prediction intervals / [prediction errors](https://opengeohub.github.io/spatial-prediction-eml/spatial-interpolation-using-ensemble-ml.html#interpolation-of-numeric-values-using-spatial-regression) for new locations (no observations) i.e. where `o$yvar` is not available?

**Comment by ishwaran** [comment]
This works for both training and testing.

If the model is 

`Y = f + noise`

then notice

`Y - f = noise`

Then using the OOB residuals we can approximate the distribution of the noise measurement error which means we can develop an interval for Y - f from the OOB residuals.    Suppose this interval is [Q_L,Q_U] obtained by taking the lower and upper quantiles of the OOB residuals at some value (say 0.05 and .95). Then to get the prediction interval for Y, use

`[Q_L + f*, Q_U + f*]`

where f* is the random forest predicted value (which can be either training or testing).  

Note all of this is based on an equal variance model assumption.  Also the prediction interval is conditional on f*.
